### PR TITLE
Add support for different and multiple durations.

### PR DESCRIPTION
**Pull request by mandyvenables**
This is a feature that adds support for different and multiple durations.

3 new tests have been added, 2 that check for valid duration and 1 that tests functionality.

### CODE REVIEW

File: lib/restrictions.js
Comment (mattcollier) [inline lib/restrictions.js:297]: it appears to me that `iso8601-duration` provides an `end` API that can be used here:

https://github.com/tolu/ISO8601-duration/blob/master/lib/index.js#L58

File: test/mocha/10-restrictions.js
Comment (mattcollier) [inline test/mocha/10-restrictions.js:91]: I think an additional test that demonstrates that test what is/is not a `DuplicateError` would help to ensure that the index is functioning as expected.  Restrictions with the same `resource`, `zone` with distinct `methodOptions.foo` values should now be allowed where they were not before.

**Review by mattcollier — CHANGES_REQUESTED** [review]

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js:67]: Updated to use `pattern` export from the `iso8601-duration` library.

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js:68]: I looked into optional chaining, but eslint does not currently support it, even when disabling it for that line. I have added a check for `methodOptions`.

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js:297]: The `end` API gives the duration in the future date (so if the duration is P30D it will give a date 30 days in future). We are looking for the previous date, so in the P30D example we are looking for 30 days ago.

Comment (mattcollier) [reply lib/restrictions.js:297]: Yes, I see that now after studying the algo more carefully.  I was imagining a case where some "acquisitionExiprationDate" was computed from `acquistion.requested (date)` and then one would simply compare determine if that date was LTE now().  Looks to me like the current algorithm is more efficient than this idea.  Marking this resolved.  TY

File: lib/restrictions.js
Comment (mattcollier) [reply lib/restrictions.js:68]: Making the eslintrc.js file looks like this will allow for optional chainging to be used: https://gist.github.com/mattcollier/3cff1c1a51138b25d5c7c24387906db0

Also, lehn just put this in: https://github.com/digitalbazaar/eslint-config-digitalbazaar/issues/48

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js]: Yes, this will be a breaking change.

I have added validation for `restriction` so it will always contain `methodOptions` and subsequently `duration`. Records that share `resource` and `zone` but have a different `duration` are now not counted as duplicates, and I do not think that is a new feature, rather how this was intended to be implemented but not written yet.

At this time, to the best of my knowledge this does not affect any of the previous data in `bedrock-resource-restriction` or any of the projects that depend on it, because the data should already be in the exact format that is now being mandated by the validator.

File: test/mocha/10-restrictions.js
Comment (mandyvenables) [reply test/mocha/10-restrictions.js]: Added new test `should not throw DuplicateError if duration is different`.

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js:68]: This solves it for eslint, however I am still running into issues using optional chaining with the test suite. I had added the same line to the eslintrc.js file, but I am still seeing `SyntaxError: Invalid or unexpected token`.

File: lib/restrictions.js
Comment (mattcollier) [inline lib/restrictions.js:26]: style: curly brace to next line.

File: lib/restrictions.js
Comment (mattcollier) [inline lib/restrictions.js:24]: style: curly brace up one line.

File: lib/restrictions.js
Comment (mattcollier) [inline lib/restrictions.js:72]: Double quotes are typically used for this instead of single quotes.  Just by convention.  It's highly likely that we have not been 100% consistent with this.

File: schemas/restrictions.js
Comment (mattcollier) [inline schemas/restrictions.js:31]: I believe the regex pattern can be applied here as well.  http://json-schema.org/understanding-json-schema/reference/regular_expressions.html

File: schemas/restrictions.js
Comment (mattcollier) [inline schemas/restrictions.js:28]: Is there a use case for zero?  Would the minimum possibly be 1?

**Review by mattcollier — CHANGES_REQUESTED** [review]
Thanks for the schema work!

File: lib/restrictions.js
Comment (mattcollier) [reply lib/restrictions.js]: I'm going to mark this as resolved primarily because it's referencing a now obsolete line of code.  I'm adding it to the office hours agenda for tomorrow to ensure we get additional eyes on the index change.

File: lib/restrictions.js
Comment (mattcollier) [reply lib/restrictions.js:68]: fixed: https://github.com/digitalbazaar/bedrock-resource-restriction/pull/23

File: schemas/restrictions.js
Comment (mandyvenables) [reply schemas/restrictions.js:28]: I cannot think of a use case for 0, bumping to 1.

File: schemas/restrictions.js
Comment (mattcollier) [inline schemas/restrictions.js:33]: TIL that RegExp actually has a `toString` API which allows one to use the export from https://github.com/tolu/ISO8601-duration/blob/master/lib/index.js#L37

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/toString

**Review by mattcollier — CHANGES_REQUESTED** [review]
Just one more effort towards DRYing up the regex.

File: schemas/restrictions.js
Comment (mandyvenables) [reply schemas/restrictions.js]: TIL as well! The `/` at the beginning and end had to be replaced, but it worked. Thanks!

File: schemas/restrictions.js
Comment (mattcollier) [inline schemas/restrictions.js:34]: `pattern` is already a RegExp correct?  And is it imperative that the slash at front/back be removed?

Comment (mattcollier) [reply schemas/restrictions.js:34]: More to the point, I don't think it's safe to do a global replace here as a regex string might have forward slashes in the middle that should not be replaced.  If the leading/trailing slashes do need to be removed we will want to use a more targeted approach.

File: schemas/restrictions.js
Comment (mattcollier) [inline schemas/restrictions.js:6]: Cool, I think this improves readability and comprehension:

```suggestion
// convert Regex to a string
const iso8601RegexString = pattern.toString();
// remove leading and trailing slashes
const iso8601Regex =
  iso8601RegexString.substring(1, iso8601RegexString.length - 1);
```

Comment (mandyvenables) [reply schemas/restrictions.js]: Updated. 

I am questioning the convention behind these updates instead of just using the plain regex. I understand large regex strings can take away from easy to read code and proper formatting, but since this is a schema I don't think that applies as much. If I were to come to this file while debugging, the code comments would be useful, but I would still need to log out what the regex actually is so I can debug that if needed. Also, since we are importing this from a third party library, there is a (low) possibility that what is in `pattern` could be updated, introducing a bug.

What is the value of importing `pattern` and converting it to the proper format, instead of just using the regex itself?

Comment (mattcollier) [reply schemas/restrictions.js]: It's DRY principles mainly.  But also it is often the case when using well maintained libraries that we get a test suite for free.  Which is the case with this module: https://github.com/tolu/ISO8601-duration/tree/master/test

Your point about using libraries is certainly a valid concern.  Presently, we trust that if the maintainers of those packages makes a change that would break our usage of the library, they would release the change as a semver **BREAKING** change.  We would not *automatically* get breaking major releases in our packages and therefore breakage would be avoided.  This is why we at DB need to pay careful attention to how we release code because we have folk building things on our stuff as well.

Now, a package maintainer *could* accidentally release a breaking change as a patch etc.  The solution for this is using package-lock files and/or pinning to specific versions.  This means that if our application works with version 1.5.1 of a dependency it will ALWAYS continue to work with v1.5.1 if we pinned to that version.  But in the common case, we are willing to risk possible breakage for the convenience of not having to constantly be updating package files in order to receive patches and new features in our dependency tree.

Comment (mandyvenables) [reply schemas/restrictions.js]: Understood, thanks for the explanation.

File: lib/restrictions.js
Comment (dlongley) [inline lib/restrictions.js:26]: The previous model was such that there could only be one restriction per resource per zone. If we're going to allow multiple restrictions per zone, then we should give restrictions IDs and check uniqueness on that basis, not based on some other parameters associated with a particular type of restriction.

File: lib/restrictions.js
Comment (dlongley) [inline lib/restrictions.js:60]: We typically don't do this type of validation here but at other levels (e.g., HTTP API). Do we need to deviate from the usual practice here?

File: lib/validator.js
Comment (dlongley) [inline lib/validator.js:19]: This seems to be a reimplementation of bedrock-validation -- any reason not to use bedrock-validation? Also, maybe we should only be doing this at a higher HTTP API level?

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js]: I have updated the code to use an id that is a hash of the zone, resource, methodOptions, and if there is a duration, the duration.

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js:59]: We are doing the validation here because this package does not have an HTTP API level.

**Comment by mattcollier** [comment]
@dlongley @mandyvenables My memory of where this PR was at is that Longley had expressed concerns related to the fundamental approach to the problem and that a design discussion was to be had.  What is the current state of things?

**Comment by mandyvenables** [comment]
@mattcollier - @dlongley had stated that instead of using the `duration` as an indexed field, we instead create an id. I have updated the code to use an id that is a hash of the `zone`, `resource`, `methodOptions`, and if there is a duration, the `duration`.

**Review by mattcollier — APPROVED** [review]
This looks good to me, but will need explicit approval from @dlongley as the package maintainer.

File: lib/restrictions.js
Comment (dmitrizagidulin) [reply lib/restrictions.js:59]: I agree, we should validate this restriction on the MRGAS level (which is currently the only thing that's calling this module).

**Review by dmitrizagidulin — APPROVED** [review]
:+1:

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js:59]: @dlongley @dmitrizagidulin So all validation should be removed at this level and moved up to the module that calls this one?

Comment (dmitrizagidulin) [reply lib/restrictions.js:59]: @mandyvenables I believe so, yes.

File: lib/restrictions.js
Comment (dlongley) [inline lib/restrictions.js:118]: So, unfortunately, creating a deterministic hash of the zone, resource, and method just recreates the same problem we had with the unique index. A unique hash on the combination of fields X, Y, Z + a unique index on that value is equivalent to a unique compound index on fields X, Y, Z.

We don't want to limit a zone to only one restriction of a certain method for a given resource. The ID should just be a random generated ID (use bnid) and we should index on that. With this PR, we're deciding that a zone can have many different restrictions on the same resource.

Side note: While we shouldn't keep this hash for the reasons above, I should also point out that if we were going this route, we would need delimiters between the three fields (X, Y, Z), to prevent accidental/intentional harmful overlap.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/restrictions.js
Comment (dlongley) [reply lib/restrictions.js]: See my comment on the hash code -- it doesn't achieve the goal in my comment here, which is to not check for uniqueness based on restriction parameters, but only based on a restriction identifier. If that restriction identifier just becomes a deterministic result of the restriction parameters, then we're checking for uniqueness based on restriction parameters when we check the ID -- just indirectly. My understanding is that we've decided that you can have any number of restrictions you want using any parameters you want -- repeated or not.

It's up to the maintainer of restrictions to decide what restrictions exist, we don't try and guess/help them not make duplicates according to some (too) limited set of information that we possess at the low level. Remember that they can create all manner of new restriction types in their extensions -- and if we wanted to try and enforce uniqueness on that basis we'd need to expose an API to allow the extensions to produce their own unique ID. Ultimately, the caller should be able to just set their own restriction ID -- and they can do whatever they want with that field (make them random or deterministic based on whatever criteria they please).

**Comment by mandyvenables** [comment]
@dlongley I have updated the code to require an id, removed the validation at this level, and updated the tests.

File: lib/restrictions.js
Comment (dlongley) [inline lib/restrictions.js:64]:
```suggestion
      `"restriction.id" is required.`,
```

Comment (dlongley) [reply lib/restrictions.js:64]: We usually just throw a TypeError for this -- so let's do that here as well.

File: lib/restrictions.js
Comment (dlongley) [inline lib/restrictions.js:74]: So this check shouldn't happen here -- it's a layer violation. The `limitOverDuration` is an extension -- one that happens to come with the core module, but an extension nonetheless.

File: lib/restrictions.js
Comment (dlongley) [reply lib/restrictions.js:121]: This still needs to be removed.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js]: Updated

File: lib/restrictions.js
Comment (mandyvenables) [reply lib/restrictions.js]: Removed.

**Comment by mattcollier** [comment]
@mandyvenables It's unclear to me if all of Longley's comments have been addressed.  If so, please re-request a review from him.

**Review by dmitrizagidulin — APPROVED** [review]
:+1:

**Pull request merged**
